// node-awe/lib/index.js
var _			= require('underscore');
var restler		= require('restler');
var fs			= require('fs');
var 

AWE = function( url, file) {

	if ( typeof(file) == 'undefined' ) file = null;

	this.url = url;
	this.filename = file;
	this.AWEId = null;
	this.attributes = {};
};

AWE.prototype.submitJob = function( callback ) {
	if ( this.filename == null || 
		 this.url == null ||
		 this.url == undefined ) {

		callback('Error: Required is missing', null);
	}

	restler.post(this.url, {
		multipart: true,
		data: {
			'upload': restler.file(this.filename, null, fs.statSync(this.filename).size)
		}
	}).on('complete', function(result, response) {
		if (result instanceof Error ) {
			callback(result, null);
		} else {
			callback(null, result);
		}
	});

}

AWE.prototype.getStatus = function( id, callback ) {
	var url = this.url + '/' + id;
	this.AWEId = id;

	restler.get(url).on('complete', function(result, response) {
		if (result instanceof Error ) {
			callback(result, null);
		} else {
			callback(null, result);
		}
	});

}

AWE.prototype.setId = function( id, callback ) {
	this.AWEId = id;
}

module.exports = module.exports.AWE = AWE;
