// node-awe/lib/index.js
var _			= require('underscore');
var request		= require('request');
var fs			= require('fs');
var 

AWE = function( url, file) {

	if ( typeof(file) == 'undefined' ) file = null;

	this.url = url;
	this.filename = file;
	this.AWEId = null;
	this.attributes = {};
};

AWE.prototype.submitJob = function( callback ) {
	if ( this.filename == null || 
		 this.url == null ||
		 this.url == undefined ) {

		callback('Error: Required is missing', null);
	}

	console.log(this.filename);

	var r = request.post(this.url, function( error, response, body ) {
		if ( error ) {
			callback( error, null);
		} else {
			if(response.statusCode == 200){
				callback( null, JSON.parse(body));
			} else {
				callback('error: '+ response.statusCode, null)
			}
		}
	
	});

	var form = r.form();
	form.append('upload', fs.createReadStream(this.filename));
	form.timeout = 10000;
}

AWE.prototype.getStatus = function( id, callback ) {
	var url = this.url + '/' + id;
	this.AWEId = id;

	request.get(url, function(error, response, body) {
		if (error) {
			callback(error, null);
		} else {
			if(response.statusCode == 200){
				callback( null, JSON.parse(body));
			} else {
				callback('error: '+ response.statusCode, null)
			}
		}
	});

}

AWE.prototype.setId = function( id, callback ) {
	this.AWEId = id;
}

module.exports = module.exports.AWE = AWE;
